### PR TITLE
Make `ValueGraph` serializable.

### PR DESCRIPTION
**Pull request by laxmorek**
### Short description 📝

This PR makes `ValueGraph` serializable.

### Implementation 👨‍💻

- Implement `Codable` protocol in `ValueGraph`'s models. 
- Add unit tests.
- Convert `class TuistGraph.Settings`  to `struct`.

### CODE REVIEW

File: Sources/TuistGraph/Models/FileElement.swift [hunk 1]
```diff
@@ -23,3 +23,42 @@ public enum FileElement: Equatable, Hashable {
...
+
+// MARK: - Codable
+
+extension FileElement {
+    private enum Kind: String, Codable {
+        case file
+        case folderReference
+    }
+
+    enum CodingKeys: String, CodingKey {
```
Comment (fortmarek) [inline Sources/TuistGraph/Models/FileElement.swift:35]: I believe you can declare `enum CodingKeys` private as well (might be worth adding it to other parts of the PR as well if that's true)

File: Sources/TuistGraph/Models/Settings.swift [hunk 2]
```diff
@@ -69,7 +69,7 @@ extension DefaultSettings {
     }
 }
 
-public class Settings: Equatable {
+public struct Settings: Equatable, Codable {
```
Comment (fortmarek) [inline Sources/TuistGraph/Models/Settings.swift:72]: neat 👌

File: Sources/TuistGraph/Models/TargetAction.swift [hunk 3]
```diff
@@ -132,3 +132,59 @@ extension Array where Element == TargetAction {
         filter { $0.order == .post }
     }
 }
+
+// MARK: - TargetAction.Script - Codable
+
+// case tool(_ path: String, _ args: [String] = [])
```
Comment (fortmarek) [inline Sources/TuistGraph/Models/TargetAction.swift:138]: this should probably be deleted ... ?

File: Sources/TuistGraph/Models/TargetDependency.swift [hunk 4]
```diff
@@ -17,3 +17,103 @@ public enum TargetDependency: Equatable, Hashable {
...
+        case .framework:
+            let path = try container.decode(AbsolutePath.self, forKey: .path)
+            self = .framework(path: path)
+        case .xcFramework:
+            let path = try container.decode(AbsolutePath.self, forKey: .path)
+            self = .xcFramework(path: path)
+        case .library:
+            let path = try container.decode(AbsolutePath.self, forKey: .path)
+            let publicHeaders = try container.decode(AbsolutePath.self, forKey: .publicHeaders)
+            let swiftModuleMap = try container.decode(AbsolutePath?.self, forKey: .swiftModuleMap)
```
Comment (fortmarek) [inline Sources/TuistGraph/Models/TargetDependency.swift:67]: I think for optionals it's more clear to use `encodeIfPresent` and `decodeIfPresent` (this same comment applies for multiple files in this PR)

**Review by fortmarek — COMMENTED** [review]
this looks pretty good - some nit comments, thanks for adding all the tests, it's appreciated 👏 I really wish that Swift supported automatic Codable conformance synthesization as making enums with associated values codable is pretty laborious process

**Review by kwridan — COMMENTED** [review]
Thanks for the hard work on this @laxmorek !

As a suggestion, would we be open to leveraging [`AutoCodable`](https://github.com/krzysztofzablocki/Sourcery/blob/master/Templates/Templates/AutoCodable.swifttemplate) from sorcery to help us better manage Codable conformance code for enums with associated types? We've been using it in some of our tooling and its been a huge time saver :)

**Comment by laxmorek** [comment]
Hey @kwridan

I believe we will be able to remove all code for enums with associated types that manages Codable conformance when `Swift 5.5` will be released. Have you heard about ["Codable synthesis for enums with associated values"](https://github.com/apple/swift-evolution/blob/main/proposals/0295-codable-synthesis-for-enums-with-associated-values.md) proposal?

**Comment by fortmarek** [comment]
Heard about this one but did not know it's already in Swift 5.5, that's great news! The custom coding/decoding should be quite temporary then and it's not worth IMHO to integrate sourcery at the moment 🙂

File: Sources/TuistGraph/Models/TargetAction.swift (hunk 3 above)
Comment (laxmorek) [reply Sources/TuistGraph/Models/TargetAction.swift:138]: right 😅

File: Sources/TuistGraph/Models/Settings.swift (hunk 2 above)
Comment (laxmorek) [reply Sources/TuistGraph/Models/Settings.swift:72]: 🤓

**Comment by laxmorek** [comment]
Hey @fortmarek I have addressed all comments. Would you take another look on my PR?

**Review by fortmarek — APPROVED** [review]
Looking good! Only the pipeline is failing, though

**Review by pepibumur — APPROVED** [review]

**Pull request merged**